spec/haskell/src/SEL4/Model/PSpace.lhs

```diff
@@ -262,11 +264,14 @@ Clear the ghost state for user pages, cnodes, and arch-specific objects within t
 >         Arch.deleteGhost ptr bits
 >         stateAssert ksASIDMapSafe "Object deletion would leave dangling PD pointers"
 
-In "deleteObjects" above, we assert "deletionIsSafe"; that is, that there are no pointers to these objects remaining elsewhere in the kernel state. Since we cannot easily check this in the Haskell model, we assume that it is always true; the assertion is strengthened during translation into Isabelle.
+In "deleteObjects" above, we make two assertions, which, when taken together, say that there are no pointers to these objects remaining elsewhere in the kernel state. Since we cannot easily check this in the Haskell model, we assume that it is always true; the assertion is strengthened during translation into Isabelle. We separate these properties into two assertions, since they are shown to be true by different means.
 
 > deletionIsSafe :: PPtr a -> Int -> KernelState -> Bool
 > deletionIsSafe _ _ _ = True
 
+> deletionIsSafe2 :: PPtr a -> Int -> KernelState -> Bool
+> deletionIsSafe2 _ _ _ = True
```

Comment:
It's probably too much to hope for, but do these assertions have any more specificity now that there's two of them? I'm worried about the `essay_version4_final(3)_final.doc` effect. Reading it as a reviewer I have no clue what they'd be, but I guess we can live with it.

Reply:
The comment just before `deletionIsSafe` does talk a little bit about their intent. I could be more precise, but it's one of those situations where I would be referencing something that happens within the Isabelle theory files, and so if the definitions change there, the comment becomes outdated.

As for why there are two asserts, again the comment before these definitions is trying to say that there are two because they're proved by different means. Maybe I could slightly strengthen the comment to say "We separate these properties into two assertions only because they are shown to be true by different means."

Reply:
The thing is, when a reviewer goes over your diff and doesn't have a good idea of what's happening or why, you're going to get comments that aren't very smart from the author's perspective. This is one of those times. I just don't get what's going on with the two asserts, and how they're shown by different means :/
It's a bit of an art to figure out whether it makes sense to put in enough info to explain that, or whether it's hopeless. My instinct says that if you added a separate assertion rather than meld it into the previous one, then it deserves a better name than `2`. Especially now that you explained the reasoning behind the `+` above, your goal is to be able to add more assertions... then it wouldn't be super nice if we had the infrastructure, but they ended up named `3`, `4` etc. That's what I mean by the `essay_version4_final(3)_final.doc` effect, but it's hard to communicate over text.

Reply:
The first assert is shown to be true in the lemma `deletionIsSafe` (terrible name...) which is in the `detype_locale'`. The second assert is shown to be true in the lemma `deletionIsSafe2_holds` in `delete_locale`. These locales aren't comparable, so there's no nice way of proving them in one go, hence their separation.

Yes I could come up with more specific names for these asserts, but the first assert seems like a big of a grab bag of properties that we want to hold for deletion. I would have simply added another conjunct instead of writing a new assert had it been straightforward to prove it all in one go. If you think I should change the name, I can. Maybe `live_detype` or something more longwinded and more ugly

Reply:
Cool, that helps me understand what's going on, thanks.
Maybe when democracy rolls over this review, they'll have more ideas. How about renaming `deletionIsSafe2` to `deletionIsSafeLiveDetype` or something like that (`_live_detype` can also work)? The first `deletionIsSafe` is historically a grab-bag already, renaming that is probably a waste of effort.

Reply:
I'd be happy with your suggestion. Maybe we'd like to make the `2` version a grab bag for other properties that are proved in the same way, in which case we're back to having no good specific name, but I suppose we can cross that bridge when we come to it

Reply:
Cool. If it comes to that, it'll either be in the first grab bag, the same locale (thus might not need a rename), or a different one which would require a new assert.

Reply:
I do share the concern that `2` is not a helpful name. Would it be future-proof-ish to postfix with the locale the second one is proved in? (I haven't reviewed that bit yet in the proofs).